x-pack/plugins/enterprise_search/server/routes/app_search/telemetry.test.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { loggingSystemMock, savedObjectsServiceMock } from 'src/core/server/mocks';
+import { MockRouter } from '../__mocks__/router.mock';
+
+import { registerTelemetryRoute } from './telemetry';
+
+jest.mock('../../collectors/app_search/telemetry', () => ({
+  incrementUICounter: jest.fn(),
+}));
+import { incrementUICounter } from '../../collectors/app_search/telemetry';
+
+/**
+ * Since these route callbacks are so thin, these serve simply as integration tests
+ * to ensure they're wired up to the collector functions correctly. Business logic
+ * is tested more thoroughly in the collectors/telemetry tests.
+ */
+describe('App Search Telemetry API', () => {
+  const mockRouter = new MockRouter({ method: 'put', payload: 'body' });
+  const mockLogger = loggingSystemMock.create().get();
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    mockRouter.createRouter();
+
+    registerTelemetryRoute({
+      router: mockRouter.router,
+      getSavedObjectsService: () => savedObjectsServiceMock.create(),
+      log: mockLogger,
+    } as any);
+  });
+
+  describe('PUT /api/app_search/telemetry', () => {
+    it('increments the saved objects counter', async () => {
+      const successResponse = { success: true };
+      (incrementUICounter as jest.Mock).mockImplementation(jest.fn(() => successResponse));
+
+      await mockRouter.callRoute({ body: { action: 'viewed', metric: 'setup_guide' } });
+
+      expect(incrementUICounter).toHaveBeenCalledWith({
+        savedObjects: expect.any(Object),
+        uiAction: 'ui_viewed',
+        metric: 'setup_guide',
+      });
+      expect(mockRouter.response.ok).toHaveBeenCalledWith({ body: successResponse });
+    });
+
+    it('throws an error when incrementing fails', async () => {
+      (incrementUICounter as jest.Mock).mockImplementation(jest.fn(() => Promise.reject('Failed')));
+
+      await mockRouter.callRoute({ body: { action: 'error', metric: 'error' } });
+
+      expect(incrementUICounter).toHaveBeenCalled();
+      expect(mockLogger.error).toHaveBeenCalled();
+      expect(mockRouter.response.internalError).toHaveBeenCalled();
```

Comment:
Optional: Assert the logged error content

```ts
expect(loggingSystemMock.collect(mockLogger).error).toMatchInlineSnapshot()
```

Reply:
My only concern about the inline snapshot is that a stack trace is generated:

```
        Array [
          Array [
            "App Search UI telemetry error: Error: Could not find Saved Objects service
            at handler (/Users/constance/Documents/kibana/x-pack/plugins/enterprise_search/server/routes/app_search/telemetry.ts:35:44)
            at MockRouter.callRoute (/Users/constance/Documents/kibana/x-pack/plugins/enterprise_search/server/routes/__mocks__/router.mock.ts:53:11)
            at Object.it (/Users/constance/Documents/kibana/x-pack/plugins/enterprise_search/server/routes/app_search/telemetry.test.ts:69:24)
            at Promise (/Users/constance/Documents/kibana/node_modules/jest-circus/build/utils.js:198:28)
            at new Promise (<anonymous>)
            at callAsyncCircusFn (/Users/constance/Documents/kibana/node_modules/jest-circus/build/utils.js:162:10)
            at _callCircusTest (/Users/constance/Documents/kibana/node_modules/jest-circus/build/run.js:205:40)
            at process._tickCallback (internal/process/next_tick.js:68:7)",
          ],
        ]
```

which of course is going to change per-machine and thus fail the snapshot 😬

Reply:
I changed this to

```
      expect(loggingSystemMock.collect(mockLogger).error[0][0]).toEqual(
        expect.stringContaining(
          'App Search UI telemetry error: Error: Could not find Saved Objects service'
        )
      );
```

for now - would def love to hear if there's a better approach however!

Reply:
As a funny aside, when I was googling this to try and find better approaches, one of the top hits I got was Josh Dover's excellent blog post 😂 https://joshdover.com/when-not-to-use-jest-snapshots/